### ISSUE TITLE
Unable to create folder

### ISSUE THREAD

**Issue by rusher81572** [minio/minio#3994]
Hello, I am trying to create a folder with my S3 application (Cloud Explorer):

Create a folder with GUI or CLI as shown:
```
java -jar dist/CloudExplorer.jar createfolder mybucket synctest
```

I get the following error:

```
Error Message:    " (Service: Amazon S3; Status Code: 400; Error Code: XMinioInvalidObjectName; Request ID: 3L137)
HTTP Status Code: 400
AWS Error Code:   XMinioInvalidObjectName
Error Type:       Client
Request ID:       3L137
```

Here is the Java Code: (Writes a simple file and then uploads it)
```
 try {
            if (folder != null) {
                FileWriter frr = new FileWriter(temp_file, true);
                BufferedWriter bfrr = new BufferedWriter(frr);
                bfrr.write("This is a place holder file to simulate a folder for Cloud Explorer only. This file can be deleted.");
                bfrr.close();
                File temp = new File(temp_file);
                if (temp.exists()) {
                    System.out.print("\n\nAttempting to create folder: " + folder);
                    put = new Put(temp.getAbsolutePath().toString(), access_key, secret_key, bucket, endpoint, folder + File.separator, false, false, false);
                    Put.terminal = true;
                    put.run();
                    System.out.print("\nCreate folder operation complete\n\n");
                }

            }
        } catch (Exception createFolder) {
        }

```

Mino server console errors
```
time="2017-03-29T01:48:07Z" level=error msg="Unable to create an object." cause="Object name invalid: testing81572#foo/" source="[object-handlers.go:479:objectAPIHandlers.PutObjectHandler()]" stack="object-api-input-checks.go:142:checkPutObjectArgs fs-v1.go:556:fsObjects.PutObject <autogenerated>:275:(*fsObjects).PutObject object-handlers.go:457:objectAPIHandlers.PutObjectHandler api-router.go:60:(objectAPIHandlers).PutObjectHandler-fm"
```

It works fine with AWS and other S3 compatible tools. Why does Minio think that the object name is invalid?

**Comment by harshavardhana** [comment]
@rusher81572 - this is not a bug this is intentional to avoid users creating directories. There is no concept of directory on object storage only a prefix.  This is how it was historically.

But recently we have relaxed this by not throwing any error but instead returning success - it will be a no-op, it wouldn't create any directory whatsoever but it wouldn't return an error as well. This was done to ensure some applications to have them working like cyber-duck, hadoop s3 connector etc - which have directory uploads.  For some reason they end up performing a HEAD on the prefix with "/" in it looking for a folder.

Empty directories are still not allowed i.e since PUT would be a no-op it won't create any namespace entry. TO have a namespace entry you need to upload an object with content with a prefix like as shown below. 

```
PUT /testbucket/my-favorite-dir/test.txt
```

If the request is in this manner the top level prefixes are automatically created.   It is not allowed because Object Storage is not POSIX and it shouldn't be treated as such - allowing objects in this manner can cause trouble for example when you have conflicting keys on your storage.  Example like this 

```
s3:///testbucket/my-favorite-dir/
s3:///testbucket/my-favorite-dir
s3:///testbucket/my-favorite-dir/1.txt
s3:///testbucket/my-favorite-dir/2.txt
```

Now on S3 these are distinct entities now if you try to copy them out for example using `s3cmd` or `aws cli` it not possible because on POSIX you cannot differentiate between a directory `my-favorite-dir/` or a file with name `my-favorite-dir` now depending on the lexical order one of them will be created first. Leading to situation if the file is copied first you won't be able to create the directory and vice versa - so your copy operation would fail to copy the other files inside the "supposedly" folder that you created. 

Now with this S3 is allowing users to make mistakes which is perhaps not elegant and we saw some projects even erroneously creating files in this manner. So in essence really it is not fully behaving like POSIX leading to mistakes which are harder to fix in future when you have millions of files.

So on Minio we choose to avoid this behavior and not allow users to make mistakes in the first place. Conceptually you don't need to acquire a namespace if you do not have an object to upload. There is no reason to replicate POSIX style behavior on object storage and vice versa as they can never be fully compatible.

Moving on we did indeed relax it so that applications can still not fail miserably and return success such that applications would eventually need to upload a file anyways - perhaps that is the reason to create a folder in the first place. It made sense for us to relax it.

**Comment by harshavardhana** [comment]
Fixed by multiple commits

```
commit 51fa4f7fe3d952cd83c96bd6b115b3f6ffc86734
Author: Harshavardhana <harsha@minio.io>
Date:   Fri Jan 20 16:33:01 2017 -0800

    Make PutObject a nop for an object which ends with "/" and size is '0' (#3603)
    
    This helps majority of S3 compatible applications while not returning
    an error upon directory create request.
    
    Fixes #2965
```

```
commit a5e60706a2a5eeb8188fd9614787083018cc102c
Author: Anis Elleuch <vadmeste@gmail.com>
Date:   Tue Mar 14 06:20:46 2017 +0100

    xl,fs: Return 404 if object ends with a separator (#3897)
    
    HEAD Object for FS and XL was returning invalid object name when
    an object name has a trailing slash separator, this PR changes the
    behavior and will always return 404 object not found, this guarantees
    a better compatibility with S3 spec.
```

Closing this as a non issue and also relevant fixes made in this regard.
